recovery/elliptics_recovery/utils/misc.py

```diff
@@ -355,25 +356,49 @@ def __init__(self, ctx, stats):
         self.result = True
         self.recovers_in_progress = 0
         self.processed_keys = 0
+        self.pool = multiprocessing.pool.ThreadPool(processes=ctx.pool_size)
 
     def run(self):
         self.start_time = time.time()
 
-        for i in xrange(self.ctx.batch_size):
-            if not self.run_one():
+        for _ in xrange(self.ctx.batch_size):
+            if not self._try_run_one():
                 break
 
         while not self.complete.is_set():
             self.complete.wait()
 
+        self.pool.close()
+        self.pool.join()
+
         speed = self.processed_keys / (time.time() - self.start_time)
         self.stats.set_counter('recovery_speed', round(speed, 2))
         self.stats.set_counter('recovers_in_progress', self.recovers_in_progress)
         return self.result
 
-    def callback(self, result, stat):
-        self.run_one()
-        last = False
+    def _try_run_one(self):
+        try:
+            with self.lock:
+                self.recovers_in_progress += 1
+            started = self.run_one()
+            if not started:
+                with self.lock:
+                    self.recovers_in_progress -= 1
+                    last = self.recovers_in_progress == 0
+                if last:
+                    self.complete.set()
+            return started
+        except:
+            log.exception("Failed to perform callback")
```

Comment:
It should be something like `failed to recover next key`

Reply:
b200ebf